"Deferred type error" in runtime

With idempotent axioms and some other there is an strange error in runtime. For example
```
{-# LANGUAGE GADTs,PolyKinds,TypeApplications,DataKinds,TypeOperators,ScopedTypeVariables #-}

import Data.Constraint
import Data.Constraint.Nat
import GHC.TypeLits

newtype GF (n :: Nat) = GF Integer deriving Show

instance KnownNat n => Num (GF n) where
  xf@(GF x) + GF y = GF $ (x+y) `mod` (natVal xf)
  xf@(GF x) - GF y = GF $ (x-y) `mod` (natVal xf)
  xf@(GF x) * GF y = GF $ (x*y) `mod` (natVal xf)
  abs = id
  signum xf@(GF x) | x==0 = xf
                   | otherwise = GF 1
  fromInteger = GF

x :: GF 5
x = GF 3

y :: GF 5
y = GF 4

foo :: (KnownNat m, KnownNat n) => GF m -> GF n -> GF (Lcm m n)
foo m@(GF x) n@(GF y) = GF $ (x*y) `mod` (lcm (natVal m) (natVal n))

bar :: (KnownNat m)  => GF m -> GF m -> GF m
bar (x :: GF m) y = foo x y - foo y x \\ Sub @() (lcmIsIdempotent @m) \\ lcmNat @m @m 

z :: GF 5
z = bar x y

main = print z
```
This code is compiled, but when it runs, it produces the following error
```
Main.EXE: Main.hs:32:31: error:
    * Could not deduce: m ~ Lcm m m
      from the context: Lcm m m ~ m
        bound by a type expected by the context:
                   Lcm m m ~ m => GF m
        at Main.hs:32:21-69
      `m' is a rigid type variable bound by
        the type signature for:
          bar :: forall (m :: Nat). KnownNat m => GF m -> GF m -> GF m
        at Main.hs:31:8
      Expected type: GF m
        Actual type: GF (Lcm m m)
    * In the second argument of `(-)', namely `foo y x'
      In the first argument of `(\\)', namely `foo x y - foo y x'
      In the first argument of `(\\)', namely
        `foo x y - foo y x \\ Sub @() (lcmIsIdempotent @m)'
    * Relevant bindings include
        y :: GF m (bound at Main.hs:32:17)
        x :: GF m (bound at Main.hs:32:6)
        bar :: GF m -> GF m -> GF m (bound at Main.hs:32:1)
(deferred type error)
```
I found only such a crutch solution
```
{-# LANGUAGE GADTs,PolyKinds,TypeApplications,DataKinds,TypeOperators,ScopedTypeVariables #-}

import Data.Constraint
import Data.Constraint.Nat
import GHC.TypeLits
import Data.Type.Equality

newtype GF (n :: Nat) = GF Integer deriving Show

instance KnownNat n => Num (GF n) where
  xf@(GF x) + GF y = GF $ (x+y) `mod` (natVal xf)
  xf@(GF x) - GF y = GF $ (x-y) `mod` (natVal xf)
  xf@(GF x) * GF y = GF $ (x*y) `mod` (natVal xf)
  abs = id
  signum xf@(GF x) | x==0 = xf
                   | otherwise = GF 1
  fromInteger = GF

x :: GF 5
x = GF 3

y :: GF 5
y = GF 4

dictToRefl :: Dict (a ~ b) -> a :~: b
dictToRefl Dict = Refl

foo :: (KnownNat m, KnownNat n) => GF m -> GF n -> GF (Lcm m n)
foo m@(GF x) n@(GF y) = GF $ (x*y) `mod` (lcm (natVal m) (natVal n))

bar :: (KnownNat m)  => GF m -> GF m -> GF m
bar (x :: GF m) y = castWith (apply Refl $ dictToRefl lcmIsIdempotent) -- gcastWith don't work
                        (foo x y - foo y x) \\ lcmNat @m @m


z :: GF 5
z = bar x y

main = print z
```
P.S.  It would be desirable to include in the documentation information on the language extension
`TypeApplications`, without which it is impossible to use such axioms as `plusCommutes`.

The fact that you're only experiencing that error at runtime is a [GHC 8.0.2-specific bug](https://ghc.haskell.org/trac/ghc/ticket/13446). If you compile that original example in GHC 8.0.1 (or 8.2.1), you'll get a proper compilation error instead:

```
$ /opt/ghc/8.0.1/bin/ghci Bug.hsGHCi, version 8.0.1: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
[1 of 1] Compiling Main             ( Bug.hs, interpreted )

Bug.hs:28:31: error:
    • Couldn't match type ‘m’ with ‘Lcm m m’
      ‘m’ is a rigid type variable bound by
        the type signature for:
          bar :: forall (m :: Nat). KnownNat m => GF m -> GF m -> GF m
        at Bug.hs:27:8
      Expected type: GF m
        Actual type: GF (Lcm m m)
    • In the second argument of ‘(-)’, namely ‘foo y x’
      In the first argument of ‘(\\)’, namely ‘foo x y - foo y x’
      In the first argument of ‘(\\)’, namely
        ‘foo x y - foo y x \\ Sub @() (lcmIsIdempotent @m)’
    • Relevant bindings include
        y :: GF m (bound at Bug.hs:28:17)
        x :: GF m (bound at Bug.hs:28:6)
        bar :: GF m -> GF m -> GF m (bound at Bug.hs:28:1)
```

If your swap the `lcmIsIdempotent` and `lcmNat` in the `bar`:
```
bar (x :: GF m) y = foo x y - foo y x \\ lcmNat @m @m \\ Sub @() (lcmIsIdempotent @m)
```
you'll get a compillation error similar to runtime error:
```
Main.hs:32:31: error:
    * Could not deduce: m ~ Lcm m m
      from the context: Lcm m m ~ m
        bound by a type expected by the context:
                   Lcm m m ~ m => GF m
        at Main.hs:32:21-85
      `m' is a rigid type variable bound by
        the type signature for:
          bar :: forall (m :: Nat). KnownNat m => GF m -> GF m -> GF m
        at Main.hs:31:8
      Expected type: GF m
        Actual type: GF (Lcm m m)
    * In the second argument of `(-)', namely `foo y x'
      In the first argument of `(\\)', namely `foo x y - foo y x'
      In the first argument of `(\\)', namely
        `foo x y - foo y x \\ lcmNat @m @m'
    * Relevant bindings include
        y :: GF m (bound at Main.hs:32:17)
        x :: GF m (bound at Main.hs:32:6)
        bar :: GF m -> GF m -> GF m (bound at Main.hs:32:1)
```

The question remains, how to use such axioms as `lcmIsIdempotent`.

--- a/.travis.yml
+++ b/.travis.yml
@@ -78,6 +78,7 @@ script:
  - cabal configure -v2 --enable-tests --enable-benchmarks # -v2 provides useful information for debugging
  - cabal build # this builds all libraries and executables (including tests/benchmarks)
  - cabal haddock
+ - cabal test
  - cabal sdist   # tests that a source-distribution can be generated
  - export SRC_TGZ=$(cabal info . | awk '{print $2 ".tar.gz";exit}') ;
    cd dist/;

--- a/constraints.cabal
+++ b/constraints.cabal
@@ -64,3 +64,15 @@ library
       Data.Constraint.Symbol
 
   ghc-options: -Wall
+
+test-suite spec
+  type: exitcode-stdio-1.0
+  default-language: Haskell2010
+  hs-source-dirs: tests
+  main-is: Spec.hs
+  other-modules: GH55Spec
+  ghc-options: -Wall -threaded -rtsopts
+  build-depends:
+    base >= 4.7 && < 5,
+    constraints,
+    hspec >= 2

--- a/src/Data/Constraint/Nat.hs
+++ b/src/Data/Constraint/Nat.hs
@@ -45,12 +45,18 @@ import Data.Proxy
 import GHC.TypeLits
 import Unsafe.Coerce
 
-type family Min :: Nat -> Nat -> Nat where
-type family Max :: Nat -> Nat -> Nat where
-type family Div :: Nat -> Nat -> Nat where
-type family Mod :: Nat -> Nat -> Nat where
-type family Gcd :: Nat -> Nat -> Nat where
-type family Lcm :: Nat -> Nat -> Nat where
+type family Min (m::Nat) (n::Nat) :: Nat where
+    Min m m = m
+type family Max (m::Nat) (n::Nat) :: Nat where
+    Max m m = m
+type family Div (m::Nat) (n::Nat) :: Nat where
+    Div m 1 = m
+type family Mod (m::Nat) (n::Nat) :: Nat where
+    Mod 0 m = 0
+type family Gcd (m::Nat) (n::Nat) :: Nat where
+    Gcd m m = m
+type family Lcm (m::Nat) (n::Nat) :: Nat where
+   Lcm m m = m
 
 type Divides n m = n ~ Gcd n m
 
@@ -225,16 +231,16 @@ lcmAssociates :: forall a b c. Dict (Lcm (Lcm a b) c ~ Lcm a (Lcm b c))
 lcmAssociates = axiom
 
 minIsIdempotent :: forall n. Dict (Min n n ~ n)
-minIsIdempotent = axiom
+minIsIdempotent = Dict
 
 maxIsIdempotent :: forall n. Dict (Max n n ~ n)
-maxIsIdempotent = axiom
+maxIsIdempotent = Dict
 
 gcdIsIdempotent :: forall n. Dict (Gcd n n ~ n)
-gcdIsIdempotent = axiom
+gcdIsIdempotent = Dict
 
 lcmIsIdempotent :: forall n. Dict (Lcm n n ~ n)
-lcmIsIdempotent = axiom
+lcmIsIdempotent = Dict
 
 minDistributesOverPlus :: forall n m o. Dict ((n + Min m o) ~ Min (n + m) (n + o))
 minDistributesOverPlus = axiom

--- a/tests/GH55Spec.hs
+++ b/tests/GH55Spec.hs
@@ -1,0 +1,46 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+module GH55Spec (main, spec) where
+
+import Data.Constraint
+import Data.Constraint.Nat
+import GHC.TypeLits
+import Test.Hspec
+
+newtype GF (n :: Nat) = GF Integer deriving (Eq, Show)
+
+instance KnownNat n => Num (GF n) where
+  xf@(GF a) + GF b = GF $ (a+b) `mod` (natVal xf)
+  xf@(GF a) - GF b = GF $ (a-b) `mod` (natVal xf)
+  xf@(GF a) * GF b = GF $ (a*b) `mod` (natVal xf)
+  abs = id
+  signum xf@(GF a) | a==0 = xf
+                   | otherwise = GF 1
+  fromInteger = GF
+
+x :: GF 5
+x = GF 3
+
+y :: GF 5
+y = GF 4
+
+foo :: (KnownNat m, KnownNat n) => GF m -> GF n -> GF (Lcm m n)
+foo m@(GF a) n@(GF b) = GF $ (a*b) `mod` (lcm (natVal m) (natVal n))
+
+bar :: (KnownNat m) => GF m -> GF m -> GF m
+bar (a :: GF m) b = foo a b - foo b a \\ Sub @() (lcmIsIdempotent @m) \\ lcmNat @m @m
+
+z :: GF 5
+z = bar x y
+
+main :: IO ()
+main = hspec spec
+
+spec :: Spec
+spec = describe "GH #55" $
+         it "should normalize Lcm m m" $
+           z `shouldBe` (GF 0 :: GF (Lcm 5 5))

--- a/tests/GH55Spec.hs
+++ b/tests/GH55Spec.hs
@@ -3,8 +3,12 @@
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE PolyKinds #-}
 {-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeOperators #-}
+
+#if __GLASGOW_HASKELL__ >= 800
+{-# LANGUAGE TypeApplications #-}
+#endif
+
 module GH55Spec (main, spec) where
 
 import Test.Hspec

--- a/tests/GH55Spec.hs
+++ b/tests/GH55Spec.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE PolyKinds #-}
@@ -6,10 +7,12 @@
 {-# LANGUAGE TypeOperators #-}
 module GH55Spec (main, spec) where
 
+import Test.Hspec
+
+#if __GLASGOW_HASKELL__ >= 800
 import Data.Constraint
 import Data.Constraint.Nat
 import GHC.TypeLits
-import Test.Hspec
 
 newtype GF (n :: Nat) = GF Integer deriving (Eq, Show)
 
@@ -37,10 +40,14 @@ bar (a :: GF m) b = foo a b - foo b a \\ Sub @() (lcmIsIdempotent @m) \\ lcmNat 
 z :: GF 5
 z = bar x y
 
-main :: IO ()
-main = hspec spec
-
 spec :: Spec
-spec = describe "GH #55" $
+spec = describe "GH #53" $
          it "should normalize Lcm m m" $
            z `shouldBe` (GF 0 :: GF (Lcm 5 5))
+#else
+spec :: Spec
+spec = return ()
+#endif
+
+main :: IO ()
+main = hspec spec

--- a/tests/Spec.hs
+++ b/tests/Spec.hs
@@ -1,0 +1,1 @@
+{-# OPTIONS_GHC -F -pgmF hspec-discover #-}

